src/components/Modal/Modal.tsx

```diff
@@ -0,0 +1,160 @@
+import React, { CSSProperties, useCallback, useEffect, useMemo } from 'react';
+import { CSSTransition } from 'react-transition-group';
+import classNames from 'classnames';
+import { ClassValue } from 'classnames/types';
+
+import ReactDOM from 'react-dom';
+import styles from './Modal.module.css';
+
+type Size = 'sm' | 'md' | 'lg';
+type Level = 'primary' | 'warning' | 'info' | 'danger';
+
+interface ModalProps {
+  children?: React.ReactNode;
+  className?: ClassValue;
+  isOpen?: boolean;
+  onClose?: () => void;
+  size?: Size;
+  style?: CSSProperties;
+}
+
+type ModalChildProps = Pick<ModalProps, 'children' | 'className'>;
+type ModalHeaderProps = Pick<ModalProps, 'children' | 'className'> & {
+  level?: Level;
+};
+
+type ModalType = React.FC<ModalProps> & {
+  Header: React.FC<ModalHeaderProps>;
+  Footer: React.FC<ModalChildProps>;
+  Body: React.FC<ModalChildProps>;
+};
+
+const Modal: ModalType = ({
+  children,
+  className: passedClassName = '',
+  isOpen = false,
+  onClose = () => {},
+  size = 'md',
+  style: passedStyle = {},
+  ...passedProps
+}) => {
+  const handleEscape = useCallback(
+    (e: KeyboardEvent) => {
+      if (isOpen && e.code === 'Escape' && onClose) {
+        onClose();
+      }
+    },
+    [isOpen, onClose]
+  );
+
+  useEffect(() => {
+    if (isOpen) {
+      window.addEventListener('keyup', handleEscape);
+    }
+
+    if (!isOpen) {
+      window.removeEventListener('keyup', handleEscape);
+    }
+
+    return () => {
+      if (isOpen) window.removeEventListener('keyup', handleEscape);
+    };
+  }, [handleEscape, isOpen]);
+
+  useEffect(() => {
+    if (isOpen) {
+      document.body.className = classNames(
+        document.body.className,
+        styles.bodyHasOpenModal
+      );
+    } else {
+      document.body.className = document.body.className
+        .replace(styles.bodyHasOpenModal, '')
+        .trim();
+    }
+  }, [isOpen]);
+
+  const modalContentStyle = useMemo(
+    () =>
+      classNames(
+        styles.modalContent,
+        {
+          [styles[`size--${size}`]]: size,
+        },
+        passedClassName
+      ),
+    [passedClassName, size]
+  );
+  return ReactDOM.createPortal(
+    <CSSTransition
+      in={isOpen}
+      unmountOnExit
+      timeout={{ enter: 0, exit: 300 }}
+      classNames={{
+        enterDone: styles.enterDone,
+        exit: styles.exit,
+      }}
+    >
+      <div
+        role="presentation"
+        {...passedProps}
+        style={passedStyle}
+        className={styles.modal}
+        onClick={onClose}
+      >
+        <div
+          role="presentation"
+          className={modalContentStyle}
+          onClick={(e) => e.stopPropagation()}
```

Comment:
There's always a chance that a consumer will have a good reason for wanting to capture clicks inside the modal, from a parent.

Instead of using `stopPropagation`, you could handle this use case with a `handleClickBackground` function.
That function would look at the click event's target, and only call `onClose` if the event target is not inside the modal.